### ISSUE TITLE
Cyclic terms handling

### ISSUE THREAD

**Issue by flexoron** [trealla-prolog/trealla#145]
```
$ swipl (8.4.0)
?- append([L|_],[a,b,c|L],L), L = [L|R].
L = [L, a, b, c|L],
R = [a, b, c|L] ;
L = [L, _A, a, b, c|L],
R = [_A, a, b, c|L] ;
L = [L, _A, _B, a, b, c|L],
R = [_A, _B, a, b, c|L] ;
L = [L, _A, _B, _C, a, b, c|L],
R = [_A, _B, _C, a, b, c|L] .
?-

$ tpl
Trealla Prolog (c) Infradig 2020-2023, v2.14.7
?- append([L|_],[a,b,c|L],L), L = [L|R].
Segmentation fault (core dumped)
```

**Comment by flexoron** [comment]
Segfault fixed (v2.14.8) but
it's still uncertain what kind of issue it really is.

R = [R|T] added
```
$ swipl (8.4.0)

?- append([L|_],[a,b,c|L],L), L = [L|R], R = [R|T]. 
L = _S1, % where
    _S1 = [_S1|_S2],
    _S2 = [_S2, a, b, c|_S1],
R = [_S2, a, b, c|_S1],
T = [a, b, c|_S1] ;
L = _S1, % where
    _S1 = [_S1|_S2],
    _S2 = [_S2, _A, a, b, c|_S1],
R = [_S2, _A, a, b, c|_S1],
T = [_A, a, b, c|_S1] ;
L = _S1, % where
    _S1 = [_S1|_S2],
    _S2 = [_S2, _A, _B, a, b, c|_S1],
R = [_S2, _A, _B, a, b, c|_S1],
T = [_A, _B, a, b, c|_S1] ;
L = _S1, % where
    _S1 = [_S1|_S2],
    _S2 = [_S2, _A, _B, _C, a, b, c|_S1],
R = [_S2, _A, _B, _C, a, b, c|_S1],
T = [_A, _B, _C, a, b, c|_S1] .

?- halt.

$ tpl (2.14.8)

?- append([L|_],[a,b,c|L],L), L = [L|R], R = [R|T].
   L = [_A|_B], R = [_A|_B], T = [a,b,c|L]
;  L = [_A|_B], R = [_A|_B], T = [_A|_B]
;  L = [_A|_B], R = [_A|_B], T = [_A|_B]
;  L = [_A|_B], R = [_A|_B], T = [_A|_B]
;  L = [_A|_B], R = [_A|_B], T = [_A|_B]
;  L = [_A|_B], R = [_A|_B], T = [_A|_B]
;  L = [_A|_B], R = [_A|_B], T = [_C,_D,_E,_F,_G,_H,a,b,c,...]
;  L = [_A|_B], R = [_A|_B], T = [_C,_D,_E,_F,_G,_H,_I,a,b,...]
;  ... .
 false.

?- halt. 
```

**Comment by infradig** [comment]
To be honest I rue the day I ever added support for cyclic terms. I don't
really think it's worth it.

**Comment by flexoron** [comment]
OK ... Placed back in the back of the head.

pmoura's explanation: https://github.com/didoudiaz/gprolog/issues/28

**Comment by infradig** [comment]
You should reopen this.

**Comment by flexoron** [comment]
```
$ ./tpl (2.14.8)  # Looks good

?- append([L|_],[a,b,c|L],L), L = [L|R], R = [R|T], T = [x,y,z|[d,f,g|_]].
   L = [_A|_B], R = [_A|_B], T = [x,y,z,d,f,g,a,b,c,...]
;  L = [_A|_B], R = [_A|_B], T = [x,y,z,d,f,g,_C,a,b,...]
;  L = [_A|_B], R = [_A|_B], T = [x,y,z,d,f,g,_C,_D,a,...]
?- halt.

$ ./tpl (2.14.9) (2.14.10) # Looks bad

?- append([L|_],[a,b,c|L],L), L = [L|R], R = [R|T], T = [x,y,z|[d,f,g|_]].
*** OOPS src/unify.c 1162
   throw(error(representation_error(term),(=)/2)).
?- halt.
```

**Comment by flexoron** [comment]
```
$ ./tpl (2.14.11) # re-established

?- append([L|_],[a,b,c|L],L), L = [L|R], R = [R|T], T = [x,y,z|[d,f,g|_]].
   L = [_A|_B], R = [_A|_B], T = [x,y,z,d,f,g,a,b,c,...]

?- append([L|_],[a,b,c|L],L), L = [L|R], R = [T|T], T = [x,y,z|[d,f,g|_]].
   L = [_A|_B], R = [[x,y,z,d,f,g,a,b,c,...],x,y,z,d,f,g,a,b,...], T = [x,y,z,d,f,g,a,b,c,...]

```

**Comment by infradig** [comment]
I'm going to have to break this again. The partial fix is incorrect and breaks something else basic.

I'll come back to this.

**Comment by flexoron** [comment]
Just another example:
```
$ swipl (8.4.0)
?- Z = ([[a|Z],X]=[X,[Y|Z]]), Z.
Z = ([[a|Z], [a|Z]]=[[a|Z], [a|Z]]),
X = [a|Z],
Y = a.

$ tpl (2.14.25)
?- Z = ([[a|Z],X]=[X,[Y|Z]]), Z.
*** OOPS src/unify.c 1158
   throw(error(representation_error(term),(=)/2)).
```

**Comment by infradig** [comment]
I hope to look at this soon.

**Comment by infradig** [comment]
Dunno where this issue is at. It no longer crashes.

**Comment by flexoron** [comment]
In the middle of nowhere.
Fixed with v2.20.1 (guessed).
